### PR TITLE
[data types] Add data types for rotation and QFT bloqs

### PR DESCRIPTION
**Pull request by fdmalone**
* Sneak in: Shorten QFixedPoint -> QFxP, add QUFxP (#632)
* Updated registers for HammingWeightCompute.

### CODE REVIEW

**Comment by fdmalone** [comment]
@tanujkhattar I wasn't sure about some of the data types for these registers. In particular I have a feeling the register i put as fixed point should actually not be (just a QUInt), but it was a little confusing with the gamma parameter (I guess there is a FP register holding this value omitted from the signature?)

File: qualtran/bloqs/arithmetic/hamming_weight.py
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/hamming_weight.py:66]: Is `QUint(1)` discouraged? This looks very verbose if we have to do this for every bloq that creates a `QUInt`

File: qualtran/bloqs/qft/qft_phase_gradient.py
Comment (tanujkhattar) [inline qualtran/bloqs/qft/qft_phase_gradient.py:61]: Ditto: Do we need the `QUInt` vs `QBit` special casing?

File: qualtran/bloqs/rotations/phase_gradient.py
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:235]: I'm confused how / why the tests pass. This definitely looks like a bug? 

`AddScaledValIntoPhaseReg` exepcts a `phase_grad` register of type `QUFxp` with `phase_bitsize` fractional bitsize and `0` word bitsize. 

This bloq is called from ` HammingWeightPhasingViaPhaseGradient.build_decompose_bloq` which expects `phase_grad` as a `QUInt(self.b_grad)` and does no explicit cast before calling 

```
        out, phase_grad = bb.add(
            AddScaledValIntoPhaseReg(
                self.bitsize.bit_length(), self.b_grad, self.exponent / 2, self.b_phase
            ),
            x=out,
            phase_grad=phase_grad,
        )
```

Do you know what's going on here?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
A few nits and a question regarding a potential bug

File: qualtran/_infra/data_types.py
Comment (fdmalone) [reply qualtran/_infra/data_types.py:187]: Sure, now means less pain later, but it's only a small amount of pain. Alternatively rather than a string we could just have sign, n_word and n_frac as attributes and could return a Fxp string for construction?

File: qualtran/_infra/data_types.py
Comment (fdmalone) [reply qualtran/_infra/data_types.py:234]: No!

File: qualtran/bloqs/arithmetic/hamming_weight.py
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/hamming_weight.py:66]: Yeah this is sort of a decision that needs to be made (see #641). QUInt(1) and QFxp(0, ...) etc can probably be size 1. I don't think QInt(1) makes sense. #641 could move a some of the boilerplate to the constructors.

File: qualtran/_infra/data_types.py
Comment (tanujkhattar) [reply qualtran/_infra/data_types.py:187]: I'd rather have a builder method using `n_word` and `n_frac` and keep the string as an attribute, especially if we plan to combine `QFxp` and `QUFxp`

File: qualtran/bloqs/rotations/phase_gradient.py
Comment (fdmalone) [reply qualtran/bloqs/rotations/phase_gradient.py:235]: Well they pass because there is no type checking on the registers currently and they both have the same bitsize (num_qubits). What should the correct type be?

For migration I'm mainly focussed on removing integer bitsizes for building registers. Once that's complete we'll start enforcing types (and not just bitsizes) are consistent % generic bloqs with QAny registers.

File: qualtran/_infra/data_types.py
Comment (mpharrigan) [reply qualtran/_infra/data_types.py:187]: it's really easy to go from the actual attributes to a string and not really easy to go the other way...

File: qualtran/bloqs/arithmetic/hamming_weight.py
Comment (mpharrigan) [reply qualtran/bloqs/arithmetic/hamming_weight.py:66]: In this particular instance, I wonder whether QAny would make more sense for a junk register, since we make no claims about what it actually is. 

For places in the library that blindly use a bitsize argument (like Add) I would prefer if we mandated that the registers are larger than 1 bit anyways; since e.g. adding a 1 bit value is a very different operation than adding n>1 bit values from the perspective of compiling and costing and algorithm for our desired architecture.

Long story short: it's a code smell if your datatypes change depending on logic like this and you should think about what your bloq is first

File: qualtran/bloqs/qft/qft_phase_gradient.py
Comment (mpharrigan) [reply qualtran/bloqs/qft/qft_phase_gradient.py:61]: first, it would be helpful to have a `Registers:` section in the docstring so I know what the "q" register is supposed to be. 

Does it make sense to do a 1-bit QFT?

File: qualtran/_infra/data_types.py
Comment (fdmalone) [reply qualtran/_infra/data_types.py:187]: yeah I think 

``` python
qfxp = QFxp(signed=True, n_word=8, n_frac=7)
fxp = Fxp(float_val, dtype=qfxp.fxp_string())
```

is easier / clearer to type than

``` python
qfxp = QFxp(dtype="fxp-s8/7")
```

File: qualtran/bloqs/arithmetic/hamming_weight.py
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/hamming_weight.py:66]: > For places in the library that blindly use a bitsize argument (like Add) I would prefer if we mandated that the registers are larger than 1 bit anyways; since e.g. adding a 1 bit value is a very different operation than adding n>1 bit values from the perspective of compiling and costing and algorithm for our desired architecture. 

I think this would result in a lot of if/else blocks and code that's not generic. I agree that junk in this case should probably just be `QAny` but I don't see why it's bad to have a 1 bit integer or a 1 bit fxp number. A 1 bit fxp unsigned integer can mean a different thing compared to a 1 bit fxp number with only 1 fractional bit; and that information can be useful in the algorithm. Even if not, I think there's value in implementing classes that work for arbitrary `bitsize`.

File: qualtran/bloqs/qft/qft_phase_gradient.py
Comment (tanujkhattar) [reply qualtran/bloqs/qft/qft_phase_gradient.py:61]: > Does it make sense to do a 1-bit QFT?

Yes, it's a hadamard gate. And when writing generic code that works for arbitrary `bitsize`, it's useful to not have special casing checks at the call site to delegate to a hadamard or a QFT separately; when the decomposition can take care of that. Costings would be correct in both the cases.

File: qualtran/_infra/data_types.py
Comment (tanujkhattar) [reply qualtran/_infra/data_types.py:187]: Okay, storing and using explicit attributes SGTM. 

Just FYI - note that `n_word` for `Fxp` means total bitsize and in our previous convention, it used to mean bitsize for the integer part. We should pick a convention, document it and keep it consistent.

Comment (fdmalone) [reply qualtran/_infra/data_types.py:187]: right, I had to look at their docs before writing it. Currently we have int_bitsize and frac_bitsize and the signed bit is hiding there depending on if it's a QUFxp or a QFxp, but I think the n_word is more natural in some way. Given all this discussion I will pull out the Fxp renaming to a separate PR.

**Comment by mpharrigan** [comment]
@tanujkhattar do you think the blockers are resolved once @fdmalone resolves the merge conflicts?

File: qualtran/bloqs/qft/qft_phase_gradient.py
Comment (tanujkhattar) [inline qualtran/bloqs/qft/qft_phase_gradient.py:58]: As discussed offline,phase gradient registers should be unsigned fixed point numbers with 0 integer bits.

Comment (fdmalone) [reply qualtran/bloqs/qft/qft_phase_gradient.py:58]: I'm tidying this up now, q and x are still unsigned ints right?

File: qualtran/bloqs/qft/qft_text_book.py
Comment (tanujkhattar) [inline qualtran/bloqs/qft/qft_text_book.py:43]: I think we should not special case on self.bitsize == 1; do we have a consensus yet?

File: qualtran/bloqs/rotations/phase_gradient.py
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:233]: This should be unsigned I think

File: qualtran/bloqs/qft/qft_text_book.py
Comment (fdmalone) [reply qualtran/bloqs/qft/qft_text_book.py:43]: sorry I prematurely pushed, I'm replacing with build_from_dtypes which will handle this.

Comment (fdmalone) [reply qualtran/bloqs/qft/qft_text_book.py:43]: Or rather we aren't special casing for QUInt / QInt (1)

File: qualtran/bloqs/qft/qft_phase_gradient.py
Comment (tanujkhattar) [reply qualtran/bloqs/qft/qft_phase_gradient.py:58]: I think keeping `q` and `x` as unsigned ints is probably fine for now. We can revisit later if needed.

**Review by tanujkhattar — APPROVED** [review]

**Pull request merged**
